### PR TITLE
Support Hybrid Scan for partitioned source data

### PR DESCRIPTION
**Pull request by sezruby**
### What is the context for this pull request?

 - **Tracking Issue**: n/a
 - **Parent Issue**: n/a
 - **Dependencies**: n/a

### What changes were proposed in this pull request?

Support Hybrid Scan for partitioned source data. 

For partitioned data, we should set `basePath` option to read partitioned columns. Otherwise, partitioned columns are not included in the loaded data. In Hybrid Scan, newly appended source files in a partitioned dataset should be loaded with `basePath` so that all partitioned columns are handled properly.

### Does this PR introduce _any_ user-facing change?

Yes.
Without this change, Hybrid Scan with an index on partitioned source doesn't work - it throws an exception during query execution.

### How was this patch tested?

Unit test

### CODE REVIEW

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala [hunk 1]
```diff
@@ -381,18 +381,42 @@ object RuleUtils {
...
+          if (location.partitionSchema.isEmpty) {
+            new InMemoryFileIndex(spark, filesAppended, Map(), None)
+          } else {
+            // Set "basePath" so that partitioned columns are also included in the output schema.
+            val basePath = location.partitionSpec.partitionColumns
+              .foldRight(location.partitionSpec.partitions.head.path) { (col, path) =>
+                if (path.getName.contains(col.name)) {
+                  path.getParent
+                } else {
+                  path
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:399]: Is `else` path a possible scenario?

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala [hunk 2]
```diff
@@ -381,18 +381,42 @@ object RuleUtils {
...
-            location = newLocation,
-            dataSchema = StructType(indexSchema.filter(baseRelation.schema.contains(_))),
-            options =
-              fsRelation.options + IndexConstants.INDEX_RELATION_IDENTIFIER)(spark)
+          if (location.partitionSchema.isEmpty) {
+            new InMemoryFileIndex(spark, filesAppended, Map(), None)
+          } else {
+            // Set "basePath" so that partitioned columns are also included in the output schema.
+            val basePath = location.partitionSpec.partitionColumns
+              .foldRight(location.partitionSpec.partitions.head.path) { (col, path) =>
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:395]: Can you add a comment to the code why `foldRight` is required here? (e.g., traversing back from the child folder, etc.) Btw, is it guaranteed that `partitionColumns` will be in that order?

Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:395]: Btw, wouldn't rootPath be same as basePath if hive-partitioning is used (`location.partitionSchema.nonEmpty`)?

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala [hunk 3]
```diff
@@ -381,18 +381,42 @@ object RuleUtils {
...
         val newLocation =
-          new InMemoryFileIndex(spark, filesAppended, Map(), None)
-        val newRelation =
-          fsRelation.copy(
-            location = newLocation,
-            dataSchema = StructType(indexSchema.filter(baseRelation.schema.contains(_))),
-            options =
-              fsRelation.options + IndexConstants.INDEX_RELATION_IDENTIFIER)(spark)
+          if (location.partitionSchema.isEmpty) {
+            new InMemoryFileIndex(spark, filesAppended, Map(), None)
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:391]: How about:
```scala
val options = if (location.partitionSchema.isEmpty) {
  Map()
} else {if (location.partitionSchema.isEmpty) {
  // Map() with basePath
}
```

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala (hunk 2 above)
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:395]: https://github.com/apache/spark/blob/d6a68e0b67ff7de58073c176dd097070e88ac831/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala#L220

partitionColumns seems be in order, though it cannot be guaranteed clearly.
But I think using "=" in the condition below is also risky..

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala (hunk 1 above)
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:399]: Seems it's not possible? null value is handled with default value - `A=a\B=__HIVE_DEFAULT_PARTITION__`

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala (hunk 2 above)
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala]: Ok let me just use the number of partition columns as a partition path always includes all partition column names and handle  possible exceptional cases(?) on demand..

File: src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala [hunk 4]
```diff
@@ -35,26 +35,31 @@ import com.microsoft.hyperspace.util.FileUtils
...
+  protected val sampleDataFormatDelete = sampleDataLocationRoot + "sample2"
+  protected val sampleDataFormatDelete2 = sampleDataLocationRoot + "sample3"
+  protected val sampleDataFormatDelete3 = sampleDataLocationRoot + "sample4"
+  protected val sampleDataFormatDelete4 = sampleDataLocationRoot + "sample5"
+  protected val sampleDataFormatBoth = sampleDataLocationRoot + "sample6"
+  protected val sampleDataFormat2Append = sampleDataLocationRoot + "sample7"
+  protected val sampleDataFormat2Delete = sampleDataLocationRoot + "sample8"
+  var hyperspace: Hyperspace = _
+  var fileFormat = ""
+  var fileFormat2 = ""
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala:51]: Do you need to create `var` for this? You can just use `"parquet"` and `"json"` in the caller directly (much easier to follow).

File: src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala [hunk 5]
```diff
@@ -35,26 +35,31 @@ import com.microsoft.hyperspace.util.FileUtils
...
+  var indexScanWithAppendFilesExpected = false
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala:52]: I didn't try hard to understand what this variable does between test suites, and what is this for?

File: src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala [hunk 6]
```diff
@@ -35,26 +35,31 @@ import com.microsoft.hyperspace.util.FileUtils
...
 
   // Creates an index with given 'df' and 'indexConfig'. Then copies the first 'appendCnt'
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala:54]: Update

File: src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala [hunk 7]
```diff
@@ -35,26 +35,31 @@ import com.microsoft.hyperspace.util.FileUtils
...
   // number of input files from 'df' and deletes the last 'deleteCnt' of the input files.
   def setupIndexAndChangeData(
-      df: DataFrame,
+      sourceFileFormat: String,
+      dataPath: String,
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala:58]: `sourceFileFormat` and `sourcePath`, or `dataFileFormat` and `dataPath`.

File: src/test/scala/com/microsoft/hyperspace/index/PartitionedDataHybridScanTest.scala [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+ */
+
+package com.microsoft.hyperspace.index
+
+import org.apache.hadoop.fs.Path
+
+import com.microsoft.hyperspace.Hyperspace
+import com.microsoft.hyperspace.util.FileUtils
+
+class PartitionedDataHybridScanTest extends HybridScanTest with HyperspaceSuite {
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/PartitionedDataHybridScanTest.scala:24]: Can we add tests where users specified "basePath"?

File: src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala [hunk 9]
```diff
@@ -252,37 +270,38 @@ class HybridScanTest extends QueryTest with HyperspaceSuite {
...
               assert(bucketSpec.numBuckets === 200)
               p
-            case p @ FileSourceScanExec(_, _, _, _, _, dataFilters, _) =>
+            case p @ FileSourceScanExec(_, _, _, partitionFilters, _, dataFilters, _) =>
               // Check filter pushed down properly.
               assert(
-                dataFilters.toString.contains(" >= 2000)") && dataFilters.toString.contains(
-                  " <= 4000)"))
+                (dataFilters.toString.contains(" >= 2000)") &&
+                  dataFilters.toString.contains(" <= 4000)")) ||
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala:277]: Once you start adding more conditions in `assert` - especially `||` testing different objects `dataFilters` and `partitionFilters`, it's hard to reason which condition is met. Can you separate these into different `assert`s?

File: src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala (hunk 4 above)
Comment (imback82) [reply src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala:51]: And you can update the sample* variables as `sampleParquetDataAppend`, `sampleJsonDataAppend`, etc., unless your goal is to test different combinations like `csv`, etc. which I don't see now.

File: src/test/scala/com/microsoft/hyperspace/index/PartitionedDataHybridScanTest.scala (hunk 8 above)
Comment (imback82) [reply src/test/scala/com/microsoft/hyperspace/index/PartitionedDataHybridScanTest.scala:24]: For example, we could have a test where an index was created with `spark.read.option("basePath", "/tmp").parquet("/tmp/a=1")`, and if a new partition, `a=2` is introduced, data under `a=2` shouldn't be read. (I believe that's the right behavior).

File: src/test/scala/com/microsoft/hyperspace/index/PartitionedDataHybridScanTest.scala [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+        appendCnt = 1,
+        deleteCnt = 1)
+      setupIndexAndChangeData(
+        fileFormat2,
+        sampleDataFormat2Delete,
+        indexConfig1.copy(indexName = "index_JsonDelete"),
+        appendCnt = 0,
+        deleteCnt = 2)
+    }
+  }
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/PartitionedDataHybridScanTest.scala:102]: Do we have a test that tests when a new partition is added? If it's combined with `HybridScanTest`, I suggest separating them out to better track.

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala [hunk 11]
```diff
@@ -381,18 +383,36 @@ object RuleUtils {
...
+            _,
+            _) =>
+        val options = if (location.partitionSchema.isEmpty) {
+          Map[String, String]()
+        } else {
+          // Retrieve basePath from a partition path.
+          val basePath = location.partitionSpec.partitionColumns
+            .foldLeft(location.partitionSpec.partitions.head.path) { (path, _) =>
+              path.getParent
+            }
```
Comment (apoorvedave1) [inline src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:398]: the foldLeft is confusing. could you change it as below:
```suggestion
          val basePath: Path = {
            var path = location.partitionSpec.partitions.head.path
            for (_ <- location.partitionSpec.partitionColumns.indices) {
              path = path.getParent
            }
            path
```

File: src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala (hunk 4 above)
Comment (sezruby) [reply src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala:51]: Actually it was for `delta`. Can we keep this variable names for tests of Delta Lake Hybrid Scan?

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala [hunk 12]
```diff
@@ -381,18 +383,36 @@ object RuleUtils {
...
+
+          // Set "basePath" so that partitioned columns are also included in the output schema.
+          Map("basePath" -> basePath.toString)
```
Comment (apoorvedave1) [inline src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:401]: could you please point me to where this key-value is being used. I couldn't find it.

File: src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala (hunk 5 above)
Comment (sezruby) [reply src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala:52]: This `var` is for the code reuse and skipping not related test. Partitioned source data doesn't support "Filter Rule + Parquet format" - scanning index data & source data using one scan node.

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala (hunk 11 above)
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:398]: I prefer FP way - not using `var` if possible.
How about this impl?
```
   val basePath =
            Iterator
              .iterate(location.partitionSpec.partitions.head.path)(_.getParent)
              .drop(location.partitionSpec.partitionColumns.length).next
```

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala (hunk 12 above)
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:401]: It's used in https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala#L221.

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala (hunk 11 above)
Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:398]: +1 for `foldLeft`

File: src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala (hunk 5 above)
Comment (imback82) [reply src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala:52]: OK, can you add some comments on how this `var` should be used? It was not intuitive why this was false for partitioned tests.

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala (hunk 11 above)
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:396]: Ok then let me keep `foldLeft` impl. Thanks!

File: src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala (hunk 4 above)
Comment (sezruby) [reply src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala]: I changed them as `val` and added some comment for this. Thanks!

File: src/test/scala/com/microsoft/hyperspace/index/PartitionedDataHybridScanTest.scala (hunk 8 above)
Comment (sezruby) [reply src/test/scala/com/microsoft/hyperspace/index/PartitionedDataHybridScanTest.scala:25]: I added 2 tests using basePath but not sure if they cover what you meant.

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala [hunk 13]
```diff
@@ -181,6 +166,21 @@ object RuleUtils {
     transformed
   }
 
+  /**
```
Comment (sezruby) [inline src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:169]: rearranged by reformating

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala [hunk 14]
```diff
@@ -381,18 +381,36 @@ object RuleUtils {
...
+            baseOutput,
+            _,
+            _) =>
+        val options = if (location.partitionSchema.isEmpty) {
+          Map[String, String]()
+        } else {
+          // Retrieve basePath from a partition path.
+          val basePath = location.partitionSpec.partitionColumns
+            .foldLeft(location.partitionSpec.partitions.head.path) { (path, _) =>
+              path.getParent
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:395]: Since we are more or less reverse engineering to compute base path, can we factor out to a function and add more descriptive comments? For example:
```scala
  private def extractBasePath(partitionSpec: PartitionSpec): Option[String] = {
    if (partitionSpec == PartitionSpec.emptySpec) {
      None
    } else {
      // For example, we could have the following in PartitionSpec:
      //   - partition columns = "col1", "col2"
      //   - partitions: "/path/col1=1/col2=1", "/path/col1=1/col2=2", etc.
      // , and going up the same number of directory levels as the number of partition columns
      // will compute the base path. Note that PartitionSpec.partitions will always contain
      // all the partitions in the path, so "partitions.head" is taken as an initial value.
      val basePath = partitionSpec.partitionColumns
        .foldLeft(partitionSpec.partitions.head.path)((path, _) => path.getParent)
      Some(basePath.toString)
    }
  }
```

Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:395]: Question: do we need to worry about the case where there are multiple base paths in `PartitioningAwareFileIndex.basePaths`?

**Review by imback82 — COMMENTED** [review]
I have few minor comments, but generally looking good to me!

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala (hunk 14 above)
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:395]: Seems we don't: https://github.com/apache/spark/blob/d6a68e0b67ff7de58073c176dd097070e88ac831/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala#L166

File: src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala [hunk 15]
```diff
@@ -35,100 +35,123 @@ import com.microsoft.hyperspace.util.FileUtils
...
+  // Introduce fileFormat to support "delta" format soon.
+  val fileFormat = "parquet"
+  val fileFormat2 = "json"
+
+  // This flag is for testing plan transformation if appended files could be load with index data
+  // scan node.
+  // For now, it's applied for a very specific case: FilterIndexRule, Parquet source format,
+  // no partitioning, no deleted files. This is false by default and will be changed to true if
+  // the test is required.
+  var indexDataScanWithAppendedFilesExpected = false
```
[the remark this answers is not in the document]
Comment (sezruby) [reply src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala]: I created a new suite but not sure it's the best. Let me revisit the test structure later for Delta lake.

File: src/test/scala/com/microsoft/hyperspace/index/PartitionedDataHybridScanTest.scala [hunk 16]
```diff
@@ -0,0 +1,259 @@
...
+          .toDF("Date", "RGUID", "Query", "imprs", "clicks")
+          .write
+          .partitionBy("Date")
+          .mode("append")
+          .format(fileFormat)
+          .save(testPath)
+
+        assert(FileUtils.getDirectorySize(new Path(testPath + "/Date=2019-10-03")) > 0)
+
+          {
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/PartitionedDataHybridScanTest.scala:233]: indentation seems off.

File: src/test/scala/com/microsoft/hyperspace/index/PartitionedDataHybridScanTest.scala [hunk 17]
```diff
@@ -0,0 +1,259 @@
...
+      val indexSourcePaths = Seq(testPath + "/Date=2017-09-03", testPath + "/Date=2018-09-03")
+      val df =
+        spark.read.format(fileFormat).option("basePath", testPath).load(indexSourcePaths: _*)
+      hyperspace.createIndex(df, IndexConfig("indexName", Seq("Date"), Seq("clicks")))
+      def filterQuery(df: DataFrame): DataFrame = {
+        df.filter(df("Date") > "2000-01-01").select(df("clicks"))
+      }
+
+      withIndex("indexName") {
+        // Check the index is applied regardless of Hybrid Scan config.
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/PartitionedDataHybridScanTest.scala:201]: We could move this below `{` similar to line 234.

File: src/test/scala/com/microsoft/hyperspace/index/HybridScanTestSuite.scala [hunk 18]
```diff
@@ -0,0 +1,636 @@
...
+import org.apache.spark.sql.execution.exchange.ShuffleExchangeExec
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.sources.DataSourceRegister
+
+import com.microsoft.hyperspace.{Hyperspace, Implicits, SampleData}
+import com.microsoft.hyperspace.index.execution.BucketUnionExec
+import com.microsoft.hyperspace.index.plans.logical.BucketUnion
+import com.microsoft.hyperspace.util.FileUtils
+
+trait HybridScanTestSuite extends QueryTest with HyperspaceSuite {
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/HybridScanTestSuite.scala:35]: Could you not change the file name in this PR? All the diffs are gone. :(

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala [hunk 19]
```diff
@@ -381,24 +381,54 @@ object RuleUtils {
...
+            baseOutput,
+            _,
+            _) =>
+        val options = extractBasePath(location.partitionSpec) match {
+          case Some(basePath) =>
+            // Set "basePath" so that partitioned columns are also included in the output schema.
+            Map("basePath" -> basePath)
+          case _ =>
+            Map[String, String]()
+        }
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:395]: ni: Can we utilize `map`, `getOrElse`, `orNull`, etc. of `Option` if applicable (as long as it doesn't hurt the readability)?
```scala
        val options = extractBasePath(location.partitionSpec)
          .map { basePath =>
            // Set "basePath" so that partitioned columns are also included in the output schema.
            Map("basePath" -> basePath)
          }
          .getOrElse(Map()) // I believe this may not need type info anymore (didn't check), although either one is fine.
```

File: src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala [hunk 20]
```diff
@@ -446,9 +326,11 @@ class HybridScanTest extends QueryTest with HyperspaceSuite {
     }
   }
 
-  test("Delete-only: join index, deleted files should be excluded from each index relation.") {
-    val df1 = spark.read.parquet(sampleParquetDataLocationDelete)
-    val df2 = spark.read.parquet(sampleParquetDataLocationDelete3)
+  test(
+    "Delete-only: join rule, deleted files should be excluded from each index data" +
+      " relation.") {
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala:331]: nit: Is this change required?

File: src/test/scala/com/microsoft/hyperspace/index/HybridScanGeneralTest.scala [hunk 21]
```diff
@@ -0,0 +1,167 @@
...
+
+    dfFromSample.write.format(fileFormat).save(sampleDataFormatAppend)
+    dfFromSample.write.format(fileFormat).save(sampleDataFormatAppend2)
+    dfFromSample.write.format(fileFormat).save(sampleDataFormatDelete)
+    dfFromSample.write.format(fileFormat).save(sampleDataFormatDelete2)
+    dfFromSample.write.format(fileFormat).save(sampleDataFormatDelete3)
+    dfFromSample.write.format(fileFormat).save(sampleDataFormatDelete4)
+    dfFromSample.write.format(fileFormat).save(sampleDataFormatBoth)
+    dfFromSample.write.format(fileFormat2).save(sampleDataFormat2Append)
+    dfFromSample.write.format(fileFormat2).save(sampleDataFormat2Delete)
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/HybridScanGeneralTest.scala:42]: Can we remove unused sample data?

File: src/test/scala/com/microsoft/hyperspace/index/HybridScanGeneralTest.scala [hunk 22]
```diff
@@ -0,0 +1,167 @@
...
+
+package com.microsoft.hyperspace.index
+
+import org.apache.hadoop.fs.Path
+import org.apache.spark.sql.DataFrame
+import org.apache.spark.sql.execution.datasources._
+
+import com.microsoft.hyperspace.Hyperspace
+
+class HybridScanGeneralTest extends HybridScanTestSuite {
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/HybridScanGeneralTest.scala:25]: Maybe `HybridScanForNonPartitionedDataSuite` and `HybridScanForPartitionedDataSuite`, if I understand the separation correctly? (A small description would also help)

**Review by imback82 — COMMENTED** [review]
LGTM (pending few minor comments), thanks @sezruby!

File: src/test/scala/com/microsoft/hyperspace/index/HybridScanGeneralTest.scala (hunk 21 above)
Comment (sezruby) [reply src/test/scala/com/microsoft/hyperspace/index/HybridScanGeneralTest.scala:42]: They are all used.. do you think it's not necessary to run the testset for both partitioned data and non-partitioend data?

Comment (imback82) [reply src/test/scala/com/microsoft/hyperspace/index/HybridScanGeneralTest.scala]: Ah sorry, I thought this suite was testing only the tests where partitioned data suite couldn't test.

**Review by imback82 — APPROVED** [review]
